tools/fork-network/src/cli.rs

```diff
@@ -1,191 +1,723 @@
+use nearcore::{load_config, open_storage, NearConfig, NightshadeRuntime, NEAR_BASE};
+use rayon::iter::{IntoParallelIterator, ParallelIterator};
+use serde::Deserialize;
+use std::collections::HashSet;
+use std::fs::File;
+use std::io::BufReader;
 use std::path::{Path, PathBuf};
+use std::sync::Arc;
 use strum::IntoEnumIterator;
 
 #[derive(clap::Parser)]
+/// If --reset, then does [`reset`] and nothing else.
+/// Snapshots `~/.near/data` into `~/.near/data/fork-snapshot/data`.
+/// Checks that hash of flat storage head is the same for every shard.
+/// Reads a block corresponding to the flat head.
+/// Reads a list of validators from `--validators`.
+/// Adds validator accounts to the genesis.
+/// Deletes all columns in `~/.near/data` other than FlatState, State, DbVersion.
+/// Makes a backup of the genesis file in `~/.near/genesis.json.backup`.
+/// Creates a new genesis:
+/// * for chain `$original_chain_id + --chain_id_suffix`
+/// * with epoch_length set to `--epoch_length`
+/// * with contents set to the state roots determined by the block corresponding to the flat head.
 pub struct ForkNetworkCommand {
+    #[clap(subcommand)]
+    command: SubCommand,
+}
+
+#[derive(clap::Subcommand)]
+enum SubCommand {
+    /// Prepares the DB for doing the modifications:
+    /// * Makes a snapshot.
+    /// * Finds and persists state roots.
+    Init(InitCmd),
+
+    /// Creates a DB snapshot, then
+    /// Updates the state to ensure every account has a full access key that is known to us.
+    AmendAccessKeys(AmendAccessKeysCmd),
+
+    /// Creates a DB snapshot, then
+    /// Reads a list of validator accounts from a file
+    /// Adds validator accounts to the state
+    /// Creates a genesis file with the new validators.
+    SetValidators(SetValidatorsCmd),
+
+    /// Drops unneeded columns.
+    Finalize(FinalizeCmd),
+
+    /// Recovers from a snapshot.
+    /// Deletes the snapshot.
+    Reset(ResetCmd),
+}
+
+#[derive(clap::Parser)]
+struct InitCmd;
+
+#[derive(clap::Parser)]
+struct FinalizeCmd;
+
+#[derive(clap::Parser)]
+struct AmendAccessKeysCmd {
+    #[arg(short, long, default_value = "100000")]
+    batch_size: u64,
+}
+
+#[derive(clap::Parser)]
+struct SetValidatorsCmd {
+    /// Path to the JSON list of [`Validator`] structs containing account id and public keys.
+    /// The path can be relative to `home_dir` or an absolute path.
+    /// Example of a valid file:
+    /// [{"account_id":"validator0", "public_key":"ed25519:7PGseFbWxvYVgZ89K1uTJKYoKetWs7BJtbyXDzfbAcqX"}]
     #[arg(short, long)]
-    pub reset: bool,
+    pub validators: PathBuf,
     #[arg(short, long, default_value = "1000")]
-    pub epoch_length: u64,
+    pub epoch_length: NumBlocks,
+    #[arg(long, default_value = "-fork")]
+    pub chain_id_suffix: String,
 }
 
+#[derive(clap::Parser)]
+struct ResetCmd;
+
+#[derive(Deserialize)]
+struct Validator {
+    account_id: AccountId,
+    public_key: PublicKey,
+    balance: Option<Balance>,
```

Comment:
nit: maybe better to name this `amount`? just to keep it consistent with `near_primitives::types::AccountInfo` which is what shows up in the genesis.json validator list, and the file expected by `neard amend-genesis`

Reply:
Done